test/aleph/http_test.clj

```diff
@@ -1464,6 +1464,162 @@
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "Desired HTTP version preference order differs from ALPN config." (ex-message result)))))
+
+      (testing "with extra HTTP versions in the ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http1 :http2])))}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "ALPN config contains more HTTP versions than desired." (ex-message result)))))
+
+      (testing "with matching ALPN config"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context (netty/ssl-client-context
+                                      (assoc test-ssl/client-ssl-context-opts
+                                             :application-protocol-config
+                                             (netty/application-protocol-config [:http2 :http1])))}})]
+          (is (= :success result))))
+
+      (testing "with no ALPN config but desiring HTTP/2"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http2 :http1]
+                        :ssl-context test-ssl/client-ssl-context}})]
+          (is (instance? ExceptionInfo result))
+          (is (= "No ALPN supplied, but requested non-HTTP/1 versions that require ALPN." (ex-message result)))))
+
+      (testing "with no ALPN config but desiring only HTTP/1"
+        (let [result (try-request-with-pool
+                      {:connection-options
+                       {:http-versions [:http1]
+                        :ssl-context test-ssl/client-ssl-context}})]
+          (is (= :success result)))))
+
+    (testing "HTTP/2 without ssl-context"
+      (let [result (try-request-with-pool
+                    {:connection-options
+                     {:http-versions [:http2]
+                      ;; We only want assert that not providing `:ssl-context` is possible but since
+                      ;; the test server uses a self-signed certificate, we cannot verify it with
+                      ;; the default one. Since all we care about is ALPN here, this is fine.
+                      :insecure? true}})]
```

Comment:
I hate this key, because it's so confusing.

`insecure?` here refers to using an insecure self-signed certificate, NOT using http or h2c.

Reply:
> refers to using an insecure self-signed certificate

Nit: Not just (any) self-signed certificates but also invalid certificates (e.g. expired ones or for different domains).